### PR TITLE
initial commit for datatable cells

### PR DESCRIPTION
**Pull request by sorja**
missing calculated and some other, current support for numbers

image
S/S 24.2.2022

### CODE REVIEW

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import React from 'react'
 
-import { AssessmentName, Col, ColType, Table } from '@meta/assessment'
+import { AssessmentName, Col, ColType, Row, Table } from '@meta/assessment'
+import { _camelCase } from '@core/utils/objects/camelize'
```
Comment (minotogna) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:4]: why not Objects.camlize ?

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx [hunk 2]
```diff
@@ -23,37 +25,47 @@ const Table: React.FC<Props> = (props) => {
...
 
+  const countryIso = useCountryIso()
   const odp = table.props.odp === true
   const secondary = false // todo: missing table.props.secondary === true
   const rowsHeader = rows.filter((row) => row.props.type === RowType.header)
   const rowsData = rows.filter((row) => row.props.type !== RowType.header)
   const tableRef = useRef<HTMLTableElement>(null)
   const displayTableExportButton = !secondary && !printView && tableRef.current != null
 
+  if (!data) return null
```
Comment (minotogna) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx:36]: why not at the Datatable level as it was?

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx [hunk 3]
```diff
@@ -23,37 +25,47 @@ const Table: React.FC<Props> = (props) => {
...
+
+  // Get headers from data
+
+  const headers = Object.keys(Object.values(data[countryIso])[0]).filter(
+    (header) => !['countryIso', 'variableName'].includes(header)
+  )
```
Comment (sorja) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx:42]: Extracting the headers

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 4]
```diff
@@ -28,12 +29,13 @@ type Props = {
...
   col: Col
+  row: Row
 }
 
 const Cell: React.FC<Props> = (props) => {
-  const { data, assessmentName, sectionName, table, disabled, rowIndex, col } = props
+  const { data, assessmentName, sectionName, table, disabled, rowIndex, col, row } = props
 
-  const datum = (data?.[rowIndex] as any)?.[col.props.index as number]
+  const datum = data[_camelCase(row.props.variableName ?? '')]?.[col.props.colName]?.raw
```
Comment (sorja) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:38]: Extracting the data

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx (hunk 2 above)
Comment (sorja) [reply client/pages/Assessment/pages/Section/components/DataTable/Table/Table.tsx:36]: typo, fixed

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx (hunk 1 above)
Comment (sorja) [reply client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:4]: typo fixed

**Comment by minotogna** [comment]
calculated cells should appear, not values .

**Comment by minotogna** [comment]
> calculated cells should appear, not values .

actually all rows should be visible, only raw data is currently migrated

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx (hunk 4 above)
Comment (minotogna) [reply client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx]: create a utils objects to extract raw data. getRawData ({countryIso, tableName, variableName, colName})

**Comment by minotogna** [comment]
use new repository to get data. I deprecated the old one

File: server/controller/assessment/getTableData.ts [hunk 5]
```diff
@@ -13,32 +14,28 @@ export const getTableData = async (
...
   if (!section) throw new Error('missing section')
 
+  const tables: Record<string, any> = {}
+  tableNames.forEach((tableName) => {
+    tables[tableName] = {}
+  })
+
   return client.tx(async (t) => {
     const assessment = await AssessmentRepository.read({ name: assessmentName }, t)
+    const cycle = assessment.cycles.find((cycle) => cycle.name === cycleName)
```
Comment (minotogna) [inline server/controller/assessment/getTableData.ts:28]: these should be passed from api and must be got from AssessmentController.getOneWithCycle

File: server/controller/assessment/getTableData.ts [hunk 6]
```diff
@@ -3,6 +3,7 @@ import { AssessmentRepository } from '@server/repository'
 import { CountryIso } from '@meta/area'
 import { AssessmentName } from '@meta/assessment'
 import { TableData } from '@meta/data'
+import { TableDataRepository } from '@server/repository/cycleData'
 
 export const getTableData = async (
```
Comment (minotogna) [inline server/controller/assessment/getTableData.ts:8]: this method should belong to CycleDataController

File: server/controller/assessment/getTableData.ts [hunk 7]
```diff
@@ -13,32 +14,28 @@ export const getTableData = async (
...
-        const { data } = await AssessmentRepository.readTableData(
-          {
-            assessment,
-            countryIso,
-            cycleName,
-            tableName,
-          },
-          t
-        )
+    return TableDataRepository.getTableData(
```
Comment (minotogna) [inline server/controller/assessment/getTableData.ts:30]: remove TableDataRepository and use CycleDataRepository

**Review by minotogna — CHANGES_REQUESTED** [review]
see my comments

File: client/pages/Assessment/pages/Section/components/DataTable/utils/index.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+export const getDatum = (data: TableData, countryIso: CountryIso, table: Table, row: Row, col: Col) => {
+  const dataTable = _getDataTable(data, countryIso, table)
+  if (!dataTable) return null
+  const { colName } = col.props
+  if (!colName) return null
+  const rowName = Objects.camelize(row.props.variableName)
+  return data[countryIso][table.props.name][rowName][colName].raw
+}
```
[the remark this answers is not in the document]
Comment (sorja) [reply client/pages/Assessment/pages/Section/components/DataTable/utils/index.ts:21]: used for table headers

File: server/controller/assessment/getTableData.ts (hunk 7 above)
Comment (sorja) [reply server/controller/assessment/getTableData.ts:30]: Rename

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 9]
```diff
@@ -1,6 +1,9 @@
 import React from 'react'
 
-import { AssessmentName, Col, ColType, Table } from '@meta/assessment'
+import { AssessmentName, Col, ColType, Row, Table } from '@meta/assessment'
+import { TableData } from '@meta/data'
+import { getDatum } from '@client/pages/Assessment/pages/Section/components/DataTable/utils'
```
[the remark this answers is not in the document]
Comment (sorja) [reply client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:5]: meta/assessment/tableData.ts:
```
export const TableData = {
  getTableData,
  getDatum,
}

```

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 10]
```diff
@@ -1,6 +1,9 @@
 import React from 'react'
 
-import { AssessmentName, Col, ColType, Table } from '@meta/assessment'
+import { AssessmentName, Col, ColType, Row, Table } from '@meta/assessment'
+import { TableData as TypeTableData } from '@meta/data'
+import { useCountryIso } from '@client/hooks'
+import { TableDatas } from '@meta/data/tableDatas'
```
Comment (minotogna) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:6]: export TableDatas from meta/data

File: meta/data/tableDatas.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import { TableData as TypeTableData } from '@meta/data/index'
+import { CountryIso } from '@meta/area'
+import { Table } from '@meta/assessment/table'
+import { Row } from '@meta/assessment/row'
+import { Col } from '@meta/assessment/col'
+
+const getTableData = (data: TypeTableData, countryIso: CountryIso, table: Table) => data[countryIso][table.props.name]
+
+const getDatum = (data: TypeTableData, countryIso: CountryIso, table: Table, row: Row, col: Col) => {
```
Comment (minotogna) [inline meta/data/tableDatas.ts:9]: use (props:{...})

File: server/controller/cycleData/getTableData.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  const { countryIso, assessmentName, cycleName, section, tableNames } = props
+  if (!section) throw new Error('missing section')
+
+  const tables: Record<string, any> = {}
+  tableNames.forEach((tableName) => {
+    tables[tableName] = {}
+  })
+
+  return client.tx(async (t) => {
+    const { assessment, cycle } = await AssessmentController.getOneWithCycle({ name: assessmentName, cycleName }, t)
```
Comment (minotogna) [inline server/controller/cycleData/getTableData.ts:27]: we cant use a controller inside a controller.
pass assessment and cycle from api.

**Review by minotogna — CHANGES_REQUESTED** [review]
remove old get data

File: client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx [hunk 13]
```diff
@@ -1,6 +1,8 @@
 import React from 'react'
 
-import { AssessmentName, Col, ColType, Table } from '@meta/assessment'
+import { AssessmentName, Col, ColType, Row, Table } from '@meta/assessment'
+import { TableData as TypeTableData, TableDatas } from '@meta/data'
```
Comment (minotogna) [inline client/pages/Assessment/pages/Section/components/DataTable/Table/Row/RowData/Cell/Cell.tsx:4]: why renaming TableData ?

File: meta/data/tableDatas.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import { TableData as TypeTableData } from '@meta/data/index'
+import { CountryIso } from '@meta/area'
+import { Table } from '@meta/assessment/table'
+import { Row } from '@meta/assessment/row'
+import { Col } from '@meta/assessment/col'
+
```
Comment (minotogna) [inline meta/data/tableDatas.ts:6]: why not to import from '@meta/assessment' ?

File: server/controller/cycleData/getTableData.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  client: BaseProtocol = DB
+): Promise<TableData> => {
+  const { countryIso, tableNames, assessment, cycle } = props
+
+  const tables: Record<string, any> = {}
+  tableNames.forEach((tableName) => {
+    tables[tableName] = {}
+  })
+
+  return client.tx(async (t) => {
```
Comment (minotogna) [inline server/controller/cycleData/getTableData.ts:23]: why opening a transaction here ?

**Review by minotogna — APPROVED** [review]

**Pull request merged**